x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/IlmPolicyDeprecationChecker.java

```diff
@@ -0,0 +1,85 @@
+import org.elasticsearch.xpack.core.ilm.LifecyclePolicyMetadata;
+import org.elasticsearch.xpack.core.ilm.Phase;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.deprecation.DeprecationInfoAction.filterChecks;
+import static org.elasticsearch.xpack.deprecation.LegacyTiersDetection.DEPRECATION_COMMON_DETAIL;
+import static org.elasticsearch.xpack.deprecation.LegacyTiersDetection.DEPRECATION_HELP_URL;
+import static org.elasticsearch.xpack.deprecation.LegacyTiersDetection.DEPRECATION_MESSAGE;
+import static org.elasticsearch.xpack.deprecation.LegacyTiersDetection.containsDeprecatedFilteredAllocationConfig;
+
+/**
+ * Checks the index templates for deprecation warnings.
+ */
+public class IlmPolicyDeprecationChecker implements ResourceDeprecationChecker {
+
+    public static final String NAME = "ilm_policies";
+    private static final List<Function<LifecyclePolicy, DeprecationIssue>> CHECKS = List.of(IlmPolicyDeprecationChecker::checkIlmPolicy);
+
+    /**
+     * @param clusterState The cluster state provided for the checker
+     * @return the name of the data streams that have violated the checks with their respective warnings.
+     */
+    @Override
+    public Map<String, List<DeprecationIssue>> check(ClusterState clusterState, DeprecationInfoAction.Request request) {
+        IndexLifecycleMetadata lifecycleMetadata = clusterState.metadata().custom(IndexLifecycleMetadata.TYPE);
+        if (lifecycleMetadata == null || lifecycleMetadata.getPolicyMetadatas().isEmpty()) {
+            return Map.of();
+        }
+        Map<String, List<DeprecationIssue>> issues = new HashMap<>();
+        for (Map.Entry<String, LifecyclePolicyMetadata> entry : lifecycleMetadata.getPolicyMetadatas().entrySet()) {
+            String name = entry.getKey();
+            LifecyclePolicyMetadata policyMetadata = entry.getValue();
+
+            List<DeprecationIssue> issuesForSinglePolicy = filterChecks(CHECKS, c -> c.apply(policyMetadata.getPolicy()));
+            if (issuesForSinglePolicy.isEmpty() == false) {
+                issues.put(name, issuesForSinglePolicy);
+            }
+        }
+        return issues.isEmpty() ? Map.of() : issues;
+    }
+
+    static DeprecationIssue checkIlmPolicy(LifecyclePolicy policy) {
+        for (Phase phase : policy.getPhases().values()) {
+            AllocateAction allocateAction = (AllocateAction) phase.getActions().get(AllocateAction.NAME);
+            if (allocateAction != null) {
+                if (containsDeprecatedFilteredAllocationConfig(allocateAction.getExclude())
+                    || containsDeprecatedFilteredAllocationConfig(allocateAction.getInclude())
+                    || containsDeprecatedFilteredAllocationConfig(allocateAction.getRequire())) {
+                    return new DeprecationIssue(
+                        DeprecationIssue.Level.WARNING,
```

Comment:
Maybe out of scope for this, but should this be a critical warning, since failure to change this means that the index could fail to allocate on Cloud?

Reply:
I was going back and forth on this. 

For cloud, I agree this is critical. For self managed, assuming a user is using kibana and UA, I think it's a warning because the user might still be using the `data` node attribute (which is also a warning).

We could play it safe, and make them all of them critical (including the node attribute), on the combination: `data: hot, warm, cold, frozen`. It would be more disruptive to users but if the scope is this narrow maybe we can be more assertive. What do you think?